File: chaincode/access-log/access_log.py
from hfc.fabric import Client as FabricClient
import json

class AccessLogChaincode:

    def __init__(self):
        self.client = FabricClient()

    def initLedger(self):
        return 'Ledger initialized'

    def createLog(self, user_id, access_method, timestamp):
        # Check if key already exists
        if self.client.get_state(user_id):
            return f'Error: Key {user_id} already exists'

        # Create new log
        log_entry = {
            'user_id': user_id,
            'access_method': access_method,
            'timestamp': timestamp
        }
        self.client.put_state(user_id, json.dumps(log_entry))
        return f'Log created for {user_id}'

    def queryLog(self, user_id):
        log_entry = self.client.get_state(user_id)
        if not log_entry:
            return f'Error: No log found for {user_id}'

        return json.loads(log_entry)

def main():
    chaincode = AccessLogChaincode()
    # Example usage
    print(chaincode.initLedger())
    print(chaincode.createLog('user1', 'manual', '2024-07-22T12:00:00Z'))
    print(chaincode.queryLog('user1'))

if __name__ == '__main__':
    main()

File: chaincode/access-log/access_log.go
package main

import (
    "encoding/json"
    "fmt"
    "github.com/hyperledger/fabric-contract-api-go/contractapi"
)

type AccessLogContract struct {
    contractapi.Contract
}

type Log struct {
    UserID        string `json:"user_id"`
    AccessMethod  string `json:"access_method"`
    Timestamp     string `json:"timestamp"`
}

func (c *AccessLogContract) InitLedger(ctx contractapi.TransactionContextInterface) error {
    return nil
}

func (c *AccessLogContract) CreateLog(ctx contractapi.TransactionContextInterface, userID string, accessMethod string, timestamp string) error {
    log := Log{
        UserID:       userID,
        AccessMethod: accessMethod,
        Timestamp:    timestamp,
    }

    logAsBytes, err := json.Marshal(log)
    if err != nil {
        return err
    }

    return ctx.GetStub().PutState(userID, logAsBytes)
}

func (c *AccessLogContract) QueryLog(ctx contractapi.TransactionContextInterface, userID string) (*Log, error) {
    logAsBytes, err := ctx.GetStub().GetState(userID)
    if err != nil {
        return nil, fmt.Errorf("failed to read from world state: %v", err)
    }
    if logAsBytes == nil {
        return nil, fmt.Errorf("the log %s does not exist", userID)
    }

    var log Log
    err = json.Unmarshal(logAsBytes, &log)
    if err != nil {
        return nil, err
    }

    return &log, nil
}

func main() {
    chaincode, err := contractapi.NewChaincode(&AccessLogContract{})
    if err != nil {
        fmt.Printf("Error creating access log chaincode: %s", err.Error())
        return
    }

    if err := chaincode.Start(); err != nil {
        fmt.Printf("Error starting access log chaincode: %s", err.Error())
    }
}
